src/nvim/eval/typval_encode.c.h

```diff
@@ -0,0 +1,657 @@
+/// @brief Scope of the main function: either nothing or `static`
+
+/// @def TYPVAL_ENCODE_NAME
+/// @brief Name of the target converter
+///
+/// After including this file it will define function
+/// `encode_vim_to_{TYPVAL_ENCODE_NAME}` with scope #TYPVAL_ENCODE_SCOPE and
+/// static function `_{TYPVAL_ENCODE_NAME}_convert_one_value`.
+
+/// @def TYPVAL_ENCODE_FIRST_ARG_TYPE
+/// @brief Type of the first argument, which will be used to return the results
+///
+/// Is expected to be a pointer type.
+
+/// @def TYPVAL_ENCODE_FIRST_ARG_NAME
+/// @brief Name of the first argument
+///
+/// This name will only be used by one of the above macros which are defined by
+/// the caller. Functions defined here do not use first argument directly.
+#ifndef NVIM_EVAL_TYPVAL_ENCODE_C_H
+#define NVIM_EVAL_TYPVAL_ENCODE_C_H
+#undef NVIM_EVAL_TYPVAL_ENCODE_C_H
+
+#include <stddef.h>
+#include <inttypes.h>
+#include <assert.h>
+
+#include "nvim/lib/kvec.h"
+#include "nvim/eval_defs.h"
+#include "nvim/eval/encode.h"
+#include "nvim/func_attr.h"
+#include "nvim/eval/typval_encode.h"
+
+static int _TYPVAL_ENCODE_CONVERT_ONE_VALUE(
+    TYPVAL_ENCODE_FIRST_ARG_TYPE TYPVAL_ENCODE_FIRST_ARG_NAME,
+    MPConvStack *const mpstack, typval_T *const tv, const int copyID,
+    const char *const objname)
+  REAL_FATTR_NONNULL_ALL REAL_FATTR_WARN_UNUSED_RESULT;
+
+/// Convert single value
+///
+/// Only scalar values are converted immediately, everything else is pushed onto
+/// the stack.
+///
+/// @param  TYPVAL_ENCODE_FIRST_ARG_NAME  First argument, defined by the
+///                                       includer. Only meaningful to macros
+///                                       defined by the includer.
+/// @param[out]  mpstack  Stack with values to convert. Values which are not
+///                       converted completely by this function (i.e.
+///                       non-scalars) are pushed here.
+/// @param  tv  Converted value.
+/// @param[in]  copyID  CopyID.
+/// @param[in]  objname  Object name, used for error reporting.
+///
+/// @return OK in case of success, FAIL in case of failure.
+static int _TYPVAL_ENCODE_CONVERT_ONE_VALUE(
+    TYPVAL_ENCODE_FIRST_ARG_TYPE TYPVAL_ENCODE_FIRST_ARG_NAME,
+    MPConvStack *const mpstack, typval_T *const tv, const int copyID,
+    const char *const objname)
+{
+  switch (tv->v_type) {
+    case VAR_STRING: {
+      TYPVAL_ENCODE_CONV_STRING(tv->vval.v_string, tv_strlen(tv));
+      break;
+    }
+    case VAR_NUMBER: {
+      TYPVAL_ENCODE_CONV_NUMBER(tv->vval.v_number);
+      break;
+    }
+    case VAR_FLOAT: {
+      TYPVAL_ENCODE_CONV_FLOAT(tv->vval.v_float);
+      break;
+    }
+    case VAR_FUNC: {
+      TYPVAL_ENCODE_CONV_FUNC_START(tv->vval.v_string, false, NULL);
+      TYPVAL_ENCODE_CONV_FUNC_BEFORE_ARGS(0);
+      TYPVAL_ENCODE_CONV_FUNC_BEFORE_SELF(-1);
+      TYPVAL_ENCODE_CONV_FUNC_END();
+      break;
+    }
+    case VAR_PARTIAL: {
+      partial_T *const pt = tv->vval.v_partial;
```

Comment:
Don't we need a self reference check here to prevent an endless loop? Example: A partial in a dictionary that is also the dictionary of the partial.